lib/sailor.js

```diff
@@ -167,9 +168,49 @@ class Sailor {
 
     run() {
         const incomingQueue = this.settings.LISTEN_MESSAGES_ON;
-        const processMessage = this.processMessage.bind(this);
+        const handler = this.processMessageAndMaybeShutdownCallback.bind(this);
         log.debug('Start listening for messages on %s', incomingQueue);
-        return this.amqpConnection.listenQueue(incomingQueue, processMessage);
+        return this.amqpConnection.listenQueue(incomingQueue, handler);
+    }
+
+    processMessageAndMaybeShutdownCallback(payload, message) {
+        return this.processMessage(payload, message)
+            .finally(result => {
+                if (!this.shutdownCallback) {
+                    // shutdown isn't scheduled, do nothing
+                    return result;
+                }
+                if (this.messagesCount === 0) {
+                    // there is no another processMessage invocation, so it's time to call shutdownCallback
+                    log.debug('About to invoke shutdownCallback');
+                    this.shutdownCallback();
+                    this.shutdownCallback = null;
+                } else {
+                    // there is another not finished processMessage invocation
+                    log.debug('No shutdownCallback since messagesCount is not zero');
+                }
+                return result;
+            });
+    }
+
+    scheduleShutdown() {
+        return co(function* gen() {
+            if (this.shutdownCallback) {
+                log.debug('scheduleShutdown – shutdown is already scheduled, do nothing');
```

Comment:
It will result in disconnect + process exit in run.js. E.g. if k8s will send signals twice with some short delay.

Reply:
Good catch.
Added test for this and fixed.